File: weather-library/src/main/java/org/rcgonzalezf/weather/openweather/network/OpenWeatherApiRequest.java
package org.rcgonzalezf.weather.openweather.network;

import androidx.annotation.VisibleForTesting;

import java.util.concurrent.Executor;
import java.util.concurrent.Executors;
import org.rcgonzalezf.weather.common.models.converter.ModelConverter;
import org.rcgonzalezf.weather.common.network.ApiRequest;
import org.rcgonzalezf.weather.openweather.OpenWeatherApiCallback;
import org.rcgonzalezf.weather.openweather.converter.OpenWeatherApiModelConverter;
import org.rcgonzalezf.weather.openweather.model.ForecastData;
import org.rcgonzalezf.weather.openweather.model.OpenWeatherCurrentData;
import org.rcgonzalezf.weather.openweather.model.OpenWeatherForecastData;

public class OpenWeatherApiRequest
    implements ApiRequest<OpenWeatherApiRequestParameters, OpenWeatherApiCallback> {

  private final String mApiKey;
  private final ModelConverter<OpenWeatherForecastData, ForecastData, OpenWeatherCurrentData> mModelConverter;
  private OpenWeatherApiRequestParameters mRequestParameters;

  public OpenWeatherApiRequest(String apiKey) {
    this(apiKey, new OpenWeatherApiModelConverter());
  }

  @VisibleForTesting
  OpenWeatherApiRequest(String apiKey,
                        ModelConverter<OpenWeatherForecastData, ForecastData, OpenWeatherCurrentData> modelConverter) {
    mApiKey = apiKey;
    mModelConverter = modelConverter;
  }

  @Override public void execute(OpenWeatherApiCallback apiCallback) {
    OpenWeatherExecutor mOpenWeatherExecutor =
        new OpenWeatherExecutor(apiCallback, getExecutor(), mApiKey);
    mOpenWeatherExecutor.setModelConverter(mModelConverter);
    mOpenWeatherExecutor.performRetrofitCall(mRequestParameters);
  }

  @Override public void addRequestParameters(OpenWeatherApiRequestParameters requestParameters) {
    mRequestParameters = requestParameters;
  }

  @VisibleForTesting Executor getExecutor() {
    return Executors.newSingleThreadExecutor();
  }
}

File: weather-library/src/main/java/org/rcgonzalezf/weather/openweather/network/OpenWeatherApiRequest.kt
package org.rcgonzalezf.weather.openweather.network

import androidx.annotation.VisibleForTesting
import org.rcgonzalezf.weather.common.models.converter.ModelConverter
import org.rcgonzalezf.weather.common.network.ApiRequest
import org.rcgonzalezf.weather.openweather.OpenWeatherApiCallback
import org.rcgonzalezf.weather.openweather.converter.OpenWeatherApiModelConverter
import org.rcgonzalezf.weather.openweather.model.ForecastData
import org.rcgonzalezf.weather.openweather.model.OpenWeatherCurrentData
import org.rcgonzalezf.weather.openweather.model.OpenWeatherForecastData
import java.util.concurrent.Executor
import java.util.concurrent.Executors

open // for Mockito
class OpenWeatherApiRequest @VisibleForTesting internal constructor(private val apiKey: String?,
                                                                    private val modelConverter: ModelConverter<OpenWeatherForecastData, ForecastData, OpenWeatherCurrentData>) : ApiRequest<OpenWeatherApiRequestParameters, OpenWeatherApiCallback> {
    private var requestParameters: OpenWeatherApiRequestParameters? = null

    constructor(apiKey: String?) : this(apiKey, OpenWeatherApiModelConverter()) {}

    override fun execute(apiCallback: OpenWeatherApiCallback) {
        val mOpenWeatherExecutor = OpenWeatherExecutor(apiCallback, executor, apiKey)
        mOpenWeatherExecutor.setModelConverter(modelConverter)
        mOpenWeatherExecutor.performRetrofitCall(requestParameters)
    }

    override fun addRequestParameters(requestParameters: OpenWeatherApiRequestParameters) {
        this.requestParameters = requestParameters
    }

    // open for Mockito
    @get:VisibleForTesting
    open val executor: Executor
        get() = Executors.newSingleThreadExecutor()
}
